R/utilities.R

```diff
@@ -509,6 +509,191 @@ ExpMean <- function(x) {
   return(log(x = mean(x = exp(x = x) - 1) + 1))
 }
 
+#' Export Seurat object for UCSC cell browser
+#'
+#' @param object Seurat object
+#' @param dir output directory path
+#' @param dataset.name name of the dataset
+#' @param vars vector of metadata fields to export
+#' @param vars.names list that defines metadata field names after the export.
+#' Should map metadata column name to export name
+#' @param reductions vector of reduction names to export
+#' @param markers.file path to file with marker genes
+#' @param cluster.field name of the metadata field containing cell cluster
+#' @param cb.dir in which dir to create UCSC cell browser content root
+#' @param port \emph{experimental} on which port to run UCSC cell browser after export
+#'
+#' @return ...
+#'
+#' @author Maximilian Haeussler, Nikolay Markov
+#'
+#' @importFrom utils browseURL
+#' @importFrom reticulate py_module_available import
+#'
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' ExportToCellbrowser(object = pbmc_small, dataset.name = "PBMC", dir = "out")
+#' }
+#'
+ExportToCellbrowser <- function(
+  object,
+  dir,
+  dataset.name,
+  vars = NULL,
```

Comment:
I'm wondering if `vars` and `vars.names` could be replaced by an optionally-named vector using the `...`, similar to how `RenameIdents` works (see https://github.com/satijalab/seurat/blob/release/3.0/R/objects.R#L2316:2341). As currently done, either all variables would have to be remapped or none would be remapped (or you'd have to order the variables so that the remapped ones wold come first).

With my proposed change, the `...` would be the metadata fields to pull, and the names would be the mapping. If no names are provided, the `names` attribute of `...` would be `NULL`. If only some are named, you could easily test this using `nchar`. This could be done like so:

```R
vars <- c(...)
names(x = vars) <- names(x = vars) %||% vars
names(x = vars) <- sapply(
  X = 1:length(x = vars),
  FUN = function(i) {
    return(ifelse(test = nchar(x = names(x = vars)[i]) > 0, yes = names(x = vars[i]), no = vars[i]))
  }
)
```

What do you think?

Reply:
Thank you for this suggestion. Currently, var names are exported to the header row of `metadata.tsv` table, where they can contain almost any character, and then they are shown in the browser website as properties of cells.
Do I get it right that with this proposed strategy we would be limited to character strings that are valid in R to be names of the arguments?

Reply:
Not necessarily, it depends on if you quote named arguments or not. R variable names have to follow certain order, and `data.frame` does weird things with names, but once you fetch the the metadata variables you can replace the column names with the names of `vars` and R won't complain. I've written a dummy function that takes the `...` I've proposed and creates a data frame with the mapped names as column names.

```R
f <- function(...) {
  vars <- c(...)
  names(x = vars) <- names(x = vars) %||% vars
  names(x = vars) <- sapply(
    X = 1:length(x = vars),
    FUN = function(i) {
      return(ifelse(
        test = nchar(x = names(x = vars)[i]) > 0,
        yes = names(x = vars)[i],
        no = vars[i]
      ))
    }
  )
  df <- data.frame(as.list(x = vars)) # Replace this with actually fetching the metadata variables
  colnames(x = df) <- names(x = vars) # This is where you set the column names of the final data frame to have any character string
  return(df)
}
```

If you play with it, you'll see that R won't object to random character strings.

```R
> f(1, '1_a' = 3, c.a = 'potato', "%$#" = 5)
  1 1_a    c.a %$#
1 1   3 potato   5
```

Unquoted argument names have to follow R's argument naming conventions. Quoted argument names, however, can contain any valid character string (I haven't found a limit yet), including escape characters (such newlines) that cell browser recognizes.